common/djangoapps/student/roles.py

```diff
@@ -60,21 +62,47 @@ def strict_role_checking():
     ACCESS_ROLES_INHERITANCE.update(OLD_ACCESS_ROLES_INHERITANCE)
 
 
+def get_role_cache_course_key(course_id=None):
+    """
+    Get the cache key for the course key.
+    """
+    default = ROLE_CACHE_UNGROUPED_COURSES_KEY
+    if not course_id:
+        return default
+    if not getattr(course_id, 'html_id', False):
+        return default
+    html_id = course_id.html_id()
+    if not html_id:
+        return default
+
+    return html_id
+
+
 class BulkRoleCache:  # lint-amnesty, pylint: disable=missing-class-docstring
     CACHE_NAMESPACE = "student.roles.BulkRoleCache"
     CACHE_KEY = 'roles_by_user'
 
     @classmethod
     def prefetch(cls, users):  # lint-amnesty, pylint: disable=missing-function-docstring
-        roles_by_user = defaultdict(set)
+        roles_by_user = {}
         get_cache(cls.CACHE_NAMESPACE)[cls.CACHE_KEY] = roles_by_user
```

Comment:
This also raises my eyebrows performance-wise, although I might just be mis-understanding it.

`roles_by_user` is a dict which is stored in the cache, and the next few lines populate the dict. Does every update to `roles_by_user` trigger cache-write? Or does the code somehow know that it needs to bulk-update the `roles_by_user` cache entry once this method finishes

Reply:
I think this works because it's a Request cache, so just an in-memory dict with no write-through to an external cache server.

Reply:
Got it, ty

Reply:
Are you referring to the existing code or to my code change here?
I have to say I was quite confused by the `prefetch` method as well - the line
`get_cache(cls.CACHE_NAMESPACE)[cls.CACHE_KEY] = roles_by_user` is a pretty weird looking way to write to a cache, especially since the RequestCache has explicit set operations.
I was wondering if this even works and if so I highly doubt that this is good practice.

However my change was only to change the datastructure in the cache for each user from a set to a dict, so I didn't want to touch this to avoid breaking anything. But I was going to ask you.

Reply:
Yeah, I asked in our python slack about it and they confirmed that the RequestCache is pretty much just a writable dict. It's still weird to me that you'd do this instead of using it's explicit setter method.

Reply:
It is very likely that this usage of RequestCache just predates RequestCache having a setter method.